From: Gerald Schaefer <geraldsc@de.ibm.com>
Subject: IBM Codestream linux-2.6.15 october2005, patch 02-28

  Description: statistics: oops on cleanup
  Symptom:     NULL pointer dereference when calling
               statistic_interface_remove() or statistic_remove()
               respectively
  Problem:     struct statistic is already gone when trying to use it
               in order access the parent interface's lock
  Solution:    use a local variable for the interface pointer
  Problem-ID:  -
  Archs:       s390-31, s390-64

Acked-by: Hannes Reinecke <hare@suse.de>

--- linux-2.5/lib/statistic.c	12 Dec 2005 22:54:27 -0000	1.5
+++ linux-2.5/lib/statistic.c	9 Mar 2006 11:18:16 -0000	1.5.2.1
@@ -417,17 +417,20 @@
 int statistic_remove(struct statistic **stat_ptr)
 {
 	unsigned long flags;
+	struct statistic_interface *interface;
 
 	if (!*stat_ptr)
 		return -EINVAL;
 
-	statistic_lock((*stat_ptr)->interface, flags);
+	interface = (*stat_ptr)->interface;
+
+	statistic_lock(interface, flags);
 	if ((*stat_ptr)->release)
 		(*stat_ptr)->release(*stat_ptr);
 	list_del(&(*stat_ptr)->list);
 	kfree(*stat_ptr);
 	*stat_ptr = NULL;
-	statistic_unlock((*stat_ptr)->interface, flags);
+	statistic_unlock(interface, flags);
 
 	return 0;
 }

